### PR TITLE
Extend compiler to support parallelFor with loop static params

### PR DESCRIPTION
**Pull request by fenglixa**
This PR will support all kinds of static loop parameters defined and compile for tekton pipelines,
pipeline results is same as kubeflow pipelines with argo.

Run compilered loop static workflow in tekton:
```
# tkn p start my-pipeline --showlog
? Value for param `my_pipe_param` of type `string`? (Default is `10`) 10
Pipelinerun started: my-pipeline-run-bz4gm
Waiting for logs to be available...
[my-in-coop1-loop-item-1-1 : my-in-coop1] op1 10 10

[my-in-coop1-loop-item-1-0 : my-in-coop1] op1 1 10

[my-in-coop2-loop-item-1-0 : my-in-coop2] op2 2

[my-in-coop2-loop-item-1-1 : my-in-coop2] op2 20

[my-out-cop : my-out-cop] 10
```

Run compilered withitem_nested workflow in tekton:
```
? Value for param `my_pipe_param` of type `string`? (Default is `10`) 10
Pipelinerun started: my-pipeline-run-mxbjx
Waiting for logs to be available...
[my-in-coop1-loop-items-0 : my-in-coop1] op1 1 10

[my-in-coop1-loop-items-1 : my-in-coop1] op1 10 10

[my-inner-inner-coop-loop-items-0 : my-inner-inner-coop] op1 1 100 10

[my-inner-inner-coop-loop-items-2 : my-inner-inner-coop] op1 1 300 10

[my-inner-inner-coop-loop-items-1 : my-inner-inner-coop] op1 1 200 10

[my-inner-inner-coop-loop-items-3 : my-inner-inner-coop] op1 10 100 10

[my-inner-inner-coop-loop-items-5 : my-inner-inner-coop] op1 10 300 10

[my-inner-inner-coop-loop-items-4 : my-inner-inner-coop] op1 10 200 10

^@[my-in-coop2-loop-items-0 : my-in-coop2] op2 2

[my-in-coop2-loop-items-1 : my-in-coop2] op2 20

[my-out-cop-loop-items-0 : my-out-cop] 10
```

### CODE REVIEW

**Comment by Tomcli** [comment]
I remember tekton doesn't have the concept of loop. If we just flatten all the loops we will lose some information in the pipeline (e.g. The UI will show it as a sequential pipeline instead of looping) .

**Comment by Tomcli** [comment]
if we are going to add loop only for static params, we need to raise errors when dynamic params are used and also update the readme to describe the limitations on Tekton. So other developers can be aware of this as well.

**Comment by fenglixa** [comment]
I am re-desgining the implement for static loop, will commit code later.

**Comment by fenglixa** [comment]
Re-desgining and Coding done, Now I think we can support all the static loop parameters defined for tekton pipelines.

/unhold

**Comment by fenglixa** [comment]
/cc  @Tomcli  @ckadner 

Coule you help review the code? Thanks!

**Comment by fenglixa** [comment]
> if we are going to add loop only for static params, we need to raise errors when dynamic params are used and also update the readme to describe the limitations on Tekton. So other developers can be aware of this as well.

Yes,  the error will thrown out now, while parallelFor was defined in dynamic params. I didn't add limitation to readme yet, since I would like to see any other workaround/solution or whether tekton will achieve this feature. #70 logged for search the clues.

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 1]
```diff
@@ -46,6 +48,53 @@ def my_pipeline(a: int = 1, b: str = "default value"):
   ```
   """
 
+  def _get_loop_task(self, task: Dict, op_name_to_for_loop_op):
```
Comment (Tomcli) [inline sdk/python/kfp_tekton/compiler/compiler.py]: Can you add a small comment on what this function will return to us?

**Comment by fenglixa** [comment]
Thanks @Tomcli, I addressed you comments, pls help check whether it's OK for you now?

**Comment by Tomcli** [comment]
@ckadner Do you have any comment on this PR? Functionality wise it looks good to me.

**Comment by ckadner** [comment]
```
error: patch failed: sdk/python/kfp_tekton/compiler/compiler.py:134
error: sdk/python/kfp_tekton/compiler/compiler.py: patch does not apply
```
/retest

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 2]
```diff
@@ -63,11 +118,12 @@ def _create_dag_templates(self, pipeline, op_transformers=None, params=None, op_
     for op in pipeline.ops.values():
       for transformer in op_transformers or []:
         transformer(op)
+
     tasks = []
     for op in pipeline.ops.values():
       tasks.extend(op_to_steps_handler(op))
 
-    return tasks
+    return tasks, root_group
```
Comment (ckadner) [inline sdk/python/kfp_tekton/compiler/compiler.py]: we should not change function signatures/return types unless we have to

please revert
```Python
     return tasks, root_group 
```
back to
```Python
     return tasks
```

...and in `_create_pipeline_workflow` on line 143 see comment below

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 3]
```diff
@@ -134,6 +190,18 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
       if op.timeout:
         task['timeout'] = '%ds' % op.timeout
 
+    # process loop parameters, keep this section in the behind of other processes, ahead of gen pipeline
+    op_name_to_for_loop_op = self._get_for_loop_ops(root_group)
```
Comment (ckadner) [inline sdk/python/kfp_tekton/compiler/compiler.py]: add this line:


```Python
root_group = pipeline.groups[0]
```

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 4]
```diff
@@ -84,7 +140,7 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
       params.append(param)
 
     # generate Tekton tasks from pipeline ops
-    tasks = self._create_dag_templates(pipeline, op_transformers, params)
+    tasks, root_group = self._create_dag_templates(pipeline, op_transformers, params)
```
Comment (ckadner) [inline sdk/python/kfp_tekton/compiler/compiler.py]: revert:
```Python
     tasks, root_group = self._create_dag_templates(pipeline, op_transformers, params) 
```
back to:
```Python
     tasks = self._create_dag_templates(pipeline, op_transformers, params) 
```

File: sdk/python/tests/compiler/compiler_tests.py [hunk 5]
```diff
@@ -99,6 +114,9 @@ def _test_pipeline_workflow(self, pipeline_function, pipeline_yaml, generate_pip
     try:
       compiler.TektonCompiler().compile(pipeline_function, compiled_yaml_file, generate_pipelinerun=generate_pipelinerun)
       with open(compiled_yaml_file, 'r') as f:
+        if 'loop' in pipeline_yaml:
```
Comment (ckadner) [inline sdk/python/tests/compiler/compiler_tests.py]: would be great if we could wrap this transformation into a `lambda` function and pass a `normalize_compiler_output` function as a parameter to `_test_pipeline_workflow`

That way the `_test_pipeline_workflow` method could remain general and other tests which need to modify the generated "YAML" in order to compare with "golden" output could also pass transformations.

File: sdk/python/tests/compiler/testdata/withitem_nested.yaml [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: my-in-coop1
+spec:
+  params:
+  - name: loop-item-param-00000001-subvar-a
```
Comment (ckadner) [inline sdk/python/tests/compiler/testdata/withitem_nested.yaml]: do we really need 8 digits?

File: sdk/python/kfp_tekton/compiler/compiler.py (hunk 3 above)
Comment (fenglixa) [reply sdk/python/kfp_tekton/compiler/compiler.py:194]: Great! Thanks.

File: sdk/python/tests/compiler/testdata/withitem_nested.yaml (hunk 6 above)
Comment (fenglixa) [reply sdk/python/tests/compiler/testdata/withitem_nested.yaml:21]: It was defined in LoopArguments of kfp.dsl._for_loop.py, not by us:
```
class LoopArguments(dsl.PipelineParam):
    ... 
    # number of characters in the code which is passed to the constructor
    NUM_CODE_CHARS = 8

```

**Comment by fenglixa** [comment]
Thanks @ckadner , I addressed your comments, could you please help check whether it's OK for you now?

File: sdk/python/tests/compiler/testdata/withitem_nested.yaml (hunk 6 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/withitem_nested.yaml:21]: ah, ok, thanks!

File: sdk/python/tests/compiler/compiler_tests.py [hunk 7]
```diff
@@ -126,14 +141,15 @@ def test_pipeline_transformers_workflow(self):
     from .testdata.pipeline_transformers import transform_pipeline
     self._test_pipeline_workflow(transform_pipeline, 'pipeline_transformers.yaml')
 
-  def _test_pipeline_workflow(self, pipeline_function, pipeline_yaml, generate_pipelinerun=False):
+  def _test_pipeline_workflow(self, pipeline_function, pipeline_yaml, generate_pipelinerun=False, normalize_compiler_output=None):
```
Comment (ckadner) [inline sdk/python/tests/compiler/compiler_tests.py:144]: hmm, maybe we should call this `normalize_compiler_output_function` to make it clear what this parameter is
and break it onto a new line if line exceeds 120 characters

File: sdk/python/tests/compiler/compiler_tests.py [hunk 8]
```diff
@@ -77,6 +78,20 @@ def test_retry_workflow(self):
     from .testdata.retry import retry_sample_pipeline
     self._test_pipeline_workflow(retry_sample_pipeline, 'retry.yaml')
 
+  def test_loop_static_workflow(self):
+    """
+    Test compiling a loop static params in workflow.
+    """
+    from .testdata.loop_static import pipeline
+    self._test_pipeline_workflow(pipeline, 'loop_static.yaml', normalize_compiler_output=lambda f: re.sub("loop-item-param-.*-subvar", "loop-item-param-subvar", f))
```
Comment (ckadner) [inline sdk/python/tests/compiler/compiler_tests.py:86]: break this line if it does not fit 120 character limit

File: sdk/python/tests/compiler/compiler_tests.py [hunk 9]
```diff
@@ -126,14 +141,15 @@ def test_pipeline_transformers_workflow(self):
...
-  def _test_pipeline_workflow(self, pipeline_function, pipeline_yaml, generate_pipelinerun=False):
+  def _test_pipeline_workflow(self, pipeline_function, pipeline_yaml, generate_pipelinerun=False, normalize_compiler_output=None):
     test_data_dir = os.path.join(os.path.dirname(__file__), 'testdata')
     golden_yaml_file = os.path.join(test_data_dir, pipeline_yaml)
     temp_dir = tempfile.mkdtemp()
     compiled_yaml_file = os.path.join(temp_dir, 'workflow.yaml')
     try:
       compiler.TektonCompiler().compile(pipeline_function, compiled_yaml_file, generate_pipelinerun=generate_pipelinerun)
       with open(compiled_yaml_file, 'r') as f:
+        f = normalize_compiler_output(f.read()) if normalize_compiler_output is not None else f
```
[the remark this answers is not in the document]
Comment (fenglixa) [reply sdk/python/tests/compiler/compiler_tests.py]: Updated, Thanks

File: sdk/python/tests/compiler/compiler_tests.py (hunk 7 above)
Comment (fenglixa) [reply sdk/python/tests/compiler/compiler_tests.py]: Updated, Thanks

File: sdk/python/tests/compiler/compiler_tests.py (hunk 8 above)
Comment (fenglixa) [reply sdk/python/tests/compiler/compiler_tests.py]: Updated, Thanks

**Comment by fenglixa** [comment]
Thanks @ckadner , I addressed your comments, since also other areas in compiler_tests.py have the line exceeds 120 chars, also irregular format in it. So I fomated it and committed the code.

**Comment by ckadner** [comment]
Hi @fenglixa -- please re-run the `test_kfp_samples_report` so your PR can be merged.

If we keep getting merge conflicts on the report I will change the report format to remove the success and failure tallies at the end of the report file to avoid the merge conflicts in the future. The totals would then only show in the console output when running the report script.

**Comment by fenglixa** [comment]
Hi @ckadner -- we may need to remove the sucess and failure check in Travis CI testing.
This PR don't include the code commited rencently, it keeps "Success: 13" when I re-run `test_kfp_samples_report`. But if I resolve the conflicts with the one in master, then Travis CI test will be failed.

**Comment by fenglixa** [comment]
I had to resolve the conflicts by keeping "Success: 13", to let this PR merged firstly.
@ckadner could you help provide /lgtm lable again? Thanks

**Comment by fenglixa** [comment]
Thanks @animeshsingh

**Comment by ckadner** [comment]
> Hi @ckadner -- we may need to remove the sucess and failure check in Travis CI testing.
> This PR don't include the code commited rencently, it keeps "Success: 13" when I re-run `test_kfp_samples_report`. But if I resolve the conflicts with the one in master, then Travis CI test will be failed.

The Travis/CI check only looks at the `test_kfp_samples_report.txt` in your branch and checks that you updated it accordingly (#81). That is different from the **_merge conflict_** which arises when another PR gets merged that also changes the same lines that show the number of files which compile successfully or fail. The only way to avoid that would be by removing the totals from the `test_kfp_samples_report.txt` file, but we do want to keep that report to quickly look up the status and progress of the compiler work.

```YAML
Success: 19
Failure: 11
Total:   30
```

**Pull request merged**